Optimized index sorting for OTel logs

felixbarny opened the pull request:
See https://github.com/elastic/elasticsearch/pull/118968/files/680a023e3fefe0d79affdcc81a0f7e146b1c847e#r1899527289

felixbarny committed:
Optimized index sorting for OTel logs

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

elasticsearchmachine commented:
Hi @felixbarny, I've created a changelog YAML for you.

felixbarny committed:
Update docs/changelog/119504.yaml

kkrik-es commented on x-pack/plugin/otel-data/src/main/resources/component-templates/logs-otel@mappings.yaml:
```diff
@@ -9,6 +9,9 @@ template:
       mode: logsdb
       sort:
         field: [ "resource.attributes.host.name", "@timestamp" ]
+        order: [ "desc", "desc" ]
```
This looks fine, I'm still wondering whether hostname should be asc - similar to _tsid. Any strong preference? @martijnvg too.

felixbarny replied:
TBF, I don't know what the impact of sorting asc vs desc is. I just copied what you did in https://github.com/elastic/elasticsearch/pull/118968.

Note that I had to remove sort.mode and sort.missing in 8c843f8 to fix a failing test. Please have another  look if the defaults are sill working fine.

kkrik-es replied:
They do, but there are versioning issues during upgrade - hence the index version guarding.

kkrik-es replied:
I'll update that one too to be consistent.

felixbarny committed:
Remove sort.mode and sort.missing to fix LogsIndexModeEnabledRestTestIT.testLogsAtSettingWithTimeSeriesOverride

I don't really get why you would want to set index.mode: time_series in logs@custom but these sort settings aren't compatible with tsdb

martijnvg reviewed (approved):
Minor comment LGTM otherwise.

felixbarny committed:
Sort host name asc

felixbarny committed:
Merge branch 'main' into otel-logs-optimized-sorting

martijnvg commented on docs/changelog/119504.yaml:
```diff
@@ -0,0 +1,5 @@
+pr: 119504
+summary: Optimized index sorting for OTel logs
+area: Data streams
+type: enhancement
```
Maybe certain queries that filter on timestamp will now slightly perform better. But that isn't a really a bug, so let's leave this as an enhancement.

felixbarny commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/119746">](https://github.com/elastic/elasticsearch/pull/119746)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 9.4.6 -->
